File: Témata/06 NLRM s obecnou kovarianční maticí (autokorelace)/Matlab/chapter06c.m
clear
close all
clc

%% Nastaveni cest a dat
addpath('..\Support'); %slozka je v nadrazenem adresari, proto ..\

load yankees.txt

year = yankees(:,1);
PCT = yankees(:,2); %winning percentage in year t
OBP = yankees(:,5); %team on-base percentage in year t
SLG = yankees(:,6); %team slugging average in year t
ERA = yankees(:,8); %team earned run average in year t

tic %pocitadlo casu behu

%% Odhad Gibbsovym vzorkovacem za predpokladu autokorelace (AR(1))
%Datove matice
 y_full = PCT;
 X_full = [ones(size(y_full)) OBP SLG ERA];

[n,k] = size(X_full); %pocet pozorovani "n" a pocet parametru k

 S1 = 25000;
 S0 = 25000+1;
 S = S0+S1;

%neinformativni priory pro beta a h (nejsou tedy uvadeny)
 nu_0 = 0;
 inv_V_0 = zeros(k); %V0^-1

 p = 1; %rad AR procesu
 c = 1;
 rho_0 = 0;
 Vrho_0 = c*eye(p);

 beta = zeros(k,S);
 h = zeros(1,S);
 rho = zeros(p,S);

%% vytvoreni zpozdenych promennych + adekvatni zkraceni puvodni rady
y = y_full(2:end,:);
y_1 = y_full(1:end-1,:);
X = X_full(2:end,:);
X_1 = X_full(1:end-1,:);


%% Gibbsuv vzorkovac
%pocatecni hodnoty
 h(1,1) = 1;
 count_post = 0; %pocet zamitnutych rho; lze pak vyuzit v ramci 
%Savage-Dickey pro vypocet integracni konstanty (pro p=1 lze i analyticky)

 fprintf('Posteriorni simulace\n')
 fprintf('0%%     50%%      100%%\n')
 fprintf('|')
 %pomocna promenna pro graficke znazorneni prubehu
 pom_step = 0.05; %0.05 = posun po 5% 
 pom_graph = pom_step;  
 
for s=2:S
    ystar = y-rho(:,s-1)*y_1;
    Xstar = X-rho(:,s-1)*X_1;
    
    V_1 = inv(h(:,s-1)*Xstar'*Xstar);
    beta_1 = V_1*(h(:,s-1)*Xstar'*ystar);
    beta(:,s)=beta_1+norm_rnd(V_1);

    nu_1 =n;
    h_1 = (1/nu_1*((ystar-Xstar*beta(:,s))'*(ystar-Xstar*beta(:,s))))^(-1);
    h(:,s)=gamm_rnd_Koop(h_1,nu_1,1);

    %vypocet a sestrojeni matice E (nahodnych slozek a jejich zpozdeni)
    eps_full = y_full-X_full*beta(:,s);
    eps = eps_full(1+p:end);
    E = eps_full(1:end-1);

    Vrho_1 = inv(inv(Vrho_0)+h(:,s)*E'*E);
    rho_1 = Vrho_1*(inv(Vrho_0)*rho_0+h(:,s)*E'*eps);

    testRho = 0; %vyber rho jen ze stacionarni oblasti
    while testRho==0
     rho(:,s) = rho_1+norm_rnd(Vrho_1);
     count_post = count_post+1;
     if abs(rho(:,s))<1
        testRho=1;
        count_post = count_post-1;
     end
    end
    
    %graficke zobrazeni prubehu po 5 %
    if s/S>=pom_graph    
     fprintf('|');
     pom_graph = pom_graph+pom_step;
    end
 end  
 fprintf('\n')

%vyhozeni prvnich S0 vzorku
beta(:,1:S0) = [];
h(:,1:S0) = [];
rho(:,1:S0) = [];

%Vypocet posteriornich charakteristik
 b_mean = mean(beta,2); %posteriorni stredni hodnota 
 h_mean = mean(h);
 rho_mean = mean(rho);
 b_var = mean(beta.^2,2)-b_mean.^2; %posteriorni rozptyl
 h_var = mean(h.^2)-h_mean.^2;
 rho_var = mean(rho.^2)-rho_mean.^2;


%KONVERGENCNI DIAGNOSTIKY - Gewekova CD pomoci funkce Geweke.m
theta = [beta' h' rho'];
res_converg = Geweke(theta);

%HPDI pro beta
 hperc=0.95; %HPDI
 HPDI_beta = (quantile(beta',[0.05,0.95]))';
 HPDI_h = quantile(h,[0.05,0.95]);
 HPDI_rho = quantile(rho,[0.05,0.95]); %jednorozmerny vektor ... neni potreba transpozice

 t = toc;

%Prezentace vysledku, grafy
fprintf('Prior and posterior means for beta (std. deviations in parentheses)\n');
fprintf('                Prior       Posterior      NSE         Geweke CD             %3.1f%% HPDI \n',hperc*100);
fprintf('   ------------------------------------\n');

figure
for ii=1:k
fprintf('Beta%1.0f           neinf.   %12.4f   %10.4f   %10.4f   [%10.4f, %10.4f]\n',[ii b_mean(ii) res_converg.NSE(ii) res_converg.CD(ii) HPDI_beta(ii,:)]);
fprintf('                   (N.A.) (%12.4f)\n', sqrt(b_var(ii)));

    subplot(3,2,ii)
    hist(beta(ii,:),50)
    xlabel(['\beta_' num2str(ii)])
end

fprintf('h               neinf.   %12.4f   %10.4f   %10.4f   [%10.4f, %10.4f]\n',[h_mean res_converg.NSE(k+1) res_converg.CD(k+1) HPDI_h]);
fprintf('                   (N.A.) (%12.4f)\n', sqrt(h_var));
    figure
    hist(h,50)
    xlabel('h')

figure
for ii=1:p
fprintf('rho_%1.0f     %12.4f   %12.4f   %10.4f   %10.4f   [%10.4f, %10.4f]\n',[ii rho_0 rho_mean(ii) res_converg.NSE(k+1+ii) res_converg.CD(k+1+ii) HPDI_rho(ii,:)]);
fprintf('           (%12.4f)   (%12.4f)\n',[sqrt(Vrho_0(ii,ii)) sqrt(rho_var(ii))]);

    subplot(3,2,ii)
    hist(rho(ii,:),50)
    xlabel(['\rho_' num2str(ii)])
end



fprintf('   ------------------------------------\n');
fprintf('Number of replications: %12.4f total \n',S-1);
fprintf('                        %12.4f discarded \n',S0-1);
fprintf('   ------------------------------------\n');

fprintf('Time elapsed (s): %12.2f \n',t);
fprintf('***************************************************************************\n\n\n');

File: 06_NLRM_s_obecnou_kovarianční_maticí-autokorelace/chapter06c.py
import sys
sys.path.insert(1, "/".join(sys.path[0].split("/")[:-1]))

from support.progress_info import progress_bar
from support.geweke import geweke
from support.norm_rnd import norm_rnd
from support.gamm_rnd_koop2 import gamm_rnd_koop2

import pandas as pd
import numpy as np

import math
import pdb
import matplotlib.pyplot as plt
from tabulate import tabulate
from numpy import transpose as t
from numpy.linalg import inv
from scipy.stats import norm

#######
# POZNAMKY:
# 1 - zmenit zobrazeni v tabulce - nazvy radku a sloupcu

# ===== 1. Priprava dat =====
# winning percentage in year t
# team on-base percentage in year t
# team slugging average in year t
# team earned run average in year t
data = pd.read_csv("yankees.tsv", sep="\t")

y_full = np.array([data.PCT]).T
x_full = np.array([np.ones(len(data.PCT)), data.OBP, data.SLG, data.ERA]).T

n, k = x_full.shape  # n - pocet pozorovani, k - pocet parametru

s_1 = 25_000
s_0 = 25_000 + 1
s = s_0 + s_1

# Neinformativni priory pro beta a h (nejsou uvedeny)
nu_0 = 0
inv_v_0 = np.zeros((k,k))

p = 1   # rad AR procesu
c = 1
rho_0 = 0
vrho_0 = c * np.eye(p)

beta = np.zeros((k, s))
h = np.zeros((1,s))
rho = np.zeros((p, s))

# Vytvoreni zpozdenych promennych + zkraceni puvodni rady
y = y_full[1:n, :]
y_1 = y_full[0:(n - 1), :]
x = x_full[1:n, :]
x_1 = x_full[0:(n - 1), :]

# ===== 2. Gibbsuv vzorkovac =====
# Pocatecni hodnoty
h[0][0] = 1
# pocet zamitnutych rho, lze pak vyuzit v ramci Savage-Dickey pro
# vypocet integracni konst. (pro p = 1 lze i analyticky)
count_post = 0
nu_1 = n

for i in range(1,s):
    y_star = y - rho[:, [i - 1]] @ y_1
    x_star = x - rho[:, [i - 1]] @ x_1

    v_1 = inv(h[0][i - 1] * x_star.T @ x_star)
    beta_1 = v_1 @ (h[0][i - 1] * x_star.T @ x_star.T)
    beta[:, [i]] = beta_1 + norm_rnd(v_1)

    h_1 = (
        1 / nu_1 * (
            (y_star - x_star @ beta[:, [i]]).T @ (y_star - x_star @ beta[:, [i]])
        )
    ) ** (- 1)
    h[0][i] = gamm_rnd_koop2(h_1, nu_1)

    # Vypocet a sestrojeni matice E (nahodnych slozek a jejich zpozdeni)
    eps_full = y_full - x_full @ beta[:, [i]]
    eps = eps_full[p:n, :]
    e = eps_full[0:(n - 1)]

    vrho_1 = inv(inv(vrho_0) + h[0][i] * e.T @ e)
    rho_1 = vrho_1 @ (inv(vrho_0) @ rho_0 + h[0][i] * e.T @ eps)

    # Vyber rho jen ze stacionarni oblasti
    test_rho = 0
    while test_rho == 0:
        rho[:, [i]] = rho_1 + norm_rnd(vrho_1)
        count_post += 1
        if abs(rho[:, [i]]) < 1:    # MOJE POZNAMKA: all() mozna?
            test_rho = 1
            count_post -= 1

    progress_bar(i, s)

# Vyhozeni prvnich s_0 vzorku
beta = beta[:, s_0:]
h = h[:, s_0:]
rho = rho[:, s_0:]

# Vypocet posteriornich charakteristik
theta = np.c_[beta.T, h.T, rho.T]
means = np.mean(theta, axis=0)
stds = np.std(theta, axis=0)

# ===== 3. Konvergencni diagnostiky =====
res_converg = geweke(theta)
hpdi = np.quantile(ksi, (0.05, 0.95), axis=0)

# ===== 4. Prezentace vysledku =====
headers = ["", "Prior", "Posterior", "NSE", "Geweke CD", "95% HPDI"]
params = ["beta_1","beta_2","beta_3","beta_4","h","alpha_1","alpha_2","alpha_3"]
table = []
for i in range(theta.shape[1]):
    table.append(
        [
            params[i],
            "",
            f"{round(means[i], 4)}\n({round(stds[i], 4)})",
            round(res_converg["nse"][i], 4),
            round(res_converg["cd"][i], 4),
            f"[{round(hpdi[0][i], 4)}; {round(hpdi[1][i], 4)}]"
        ]
    )

for i in range(beta_0.shape[0]):
    table[i][1] = f"{round(beta_0[i][0], 4)}\n({round(math.sqrt(var_beta_0[i][0]), 4)})"

table[4][1] = f"{round(h_0, 4)}\n({round(math.sqrt(var_h_0),4)})"
for i in range(5,8):
    table[i][1] = f"neinf.\n(N.A.)"

print("\nApriorni a posteriorni parametry:")
print(tabulate(table, headers, tablefmt="pretty"))
print(
    f"Pocet replikaci:"\
    f"\n - {s} celkem"\
    f"\n - {s_0} vyrazeno"\
    f"\nMira akceptace: {round(accept_ratio_rw, 4)}"\
    f"\nChyb: {error_check}"
)
